### ISSUE TITLE
Include query string in SEOmatic cache key

### ISSUE THREAD

**Issue by benjamin-smith** [nystudio107/craft-seomatic#325]
It appears that for SEOmatic's internal cache, the query string is not a part of the cache key https://github.com/nystudio107/craft-seomatic/blob/d5552c7/src/services/MetaContainers.php#L166

It would be great to be able to set/alter metadata tags based on QS params. For example, if a filter is applied via QS to a landing page, I may want to change the `robots` tag for that page. Or I may want to update the canonical URL.

```
{% if craft.app.request.getParam('filterName') | length > 0 %}
    {% do seomatic.tag.get("robots").content("noindex, nofollow") %}
{% endif %}
```

**Comment by khalwat** [comment]
Yeah the reason the query string isn't part of the internal cache key is because it can easily cause the cache to balloon out of control in terms of size... especially on sites that are doing heavy referral/advertising traffic.

If you really want to do what you're suggesting, you can do something like:

```
{% set tagContainer = seomatic.tag.container() %}
{% do tagContainer.clearCache(true) %}
```

...which will disable the level 2 cache for that container. There will be a performance penalty, though.

**Comment by khalwat** [comment]
Feel free to respond here or ping me on Discord if you think it merits further discussion.

**Comment by benjamin-smith** [comment]
Thanks for the quick response! I realize this may be a separate issue, but I want to make sure I understand your workaround before I go that route.

The following code does _not_ clear the cache on each request for me. If I first hit a URL w/ `?filterName=foo`, a subsequent request to the same URL without the QS shows the same value for the `robots` meta tag.

Manually clearing Craft's caches (via CP or console) does work, and then then next request is not cached.

```
{% set tagContainer = seomatic.tag.container() %}
{% do tagContainer.clearCache(true) %}

{% if craft.app.request.getParam('filterName') | length > 0 %}
    {% do seomatic.tag.get("robots").content("noindex, nofollow") %}
{% endif %}
```

Am I understanding your workaround, or am I missing it? Note I understand the performance issue here, assuming I can get it working.

**Comment by khalwat** [comment]
Are there any other layers of caching involved here? The code you showed should indeed clear the level 2 cache (which is the parsed result of the tags):

https://github.com/nystudio107/craft-seomatic/blob/v3/src/models/MetaTagContainer.php#L52

Are you able to test this in an environment with `devMode` on (which will reduce the cache time to just 30 seconds) to see if the code path goes where it should?

**Comment by benjamin-smith** [comment]
I have confirmed there is no other caching in place, and `devMode` is on. I am seeing the cache clear after 30 seconds. I'll keep poking around to see what might be the cause of the issue.

Separately, would you be open to reviewing a (currently non-existent) PR that would allow for some sort of config variable that allows for either manually specifying cache keys or some other config setting to optionally allow for including the QS vars as part of the cache key?

**Comment by khalwat** [comment]
ooooh, I think I see a bug:

```php
        $cache = Craft::$app->getCache();
        if ($this->clearCache) {
            TagDependency::invalidate($cache, $uniqueKey);
        }
        $tagData = $cache->getOrSet(
            $this::CONTAINER_TYPE.$uniqueKey,
            function () use ($uniqueKey) {
                Craft::info(
                    $this::CONTAINER_TYPE.' cache miss: '.$uniqueKey,
                    __METHOD__
                );
```

We invalidate the TagDependency based on `$uniqueKey` but pass in `$this::CONTAINER_TYPE.$uniqueKey` as the actual key to use.

d'oh!

**Comment by khalwat** [comment]
Fixed in https://github.com/nystudio107/craft-seomatic/commit/bae706ce032b6589039689d560769d028404cf61

Can you set your semver to `dev-develop` to give it a whirl @benjamin-smith ?

**Comment by khalwat** [comment]
But yes, I'd be open to the idea of a way to include the QS in the cache key, for people who are really super sure that they want to do that (ie, it'd be deeply hidden :) )

It really can end up being rather disastrous in many circumstances.

**Comment by benjamin-smith** [comment]
Not ignoring you - it doesn't appear to fix my issue, but I am trying to debug further before I come back to you!

**Comment by khalwat** [comment]
Color me very surprised. Let me know what you uncover.

**Comment by benjamin-smith** [comment]
It looks like this call to `MetaContainers::loadMetaContainers()` is being cached on each request. The Twig tags that are set up for clearing the cache have no effect here from what I can tell. Here is how that cache key is set: (https://github.com/nystudio107/craft-seomatic/blob/d5552c7/src/services/MetaContainers.php#L191)

I am seeing how the [logic you implemented](https://github.com/nystudio107/craft-seomatic/commit/bae706ce032b6589039689d560769d028404cf61) should be working, but I cannot reproduce due to the above, which fires first. Because that cache `getOrSet()` call is higher up in the stack, the code from your latest change is never triggered.

I temporarily added `TagDependency::invalidate($cache, $this::CACHE_KEY.$cacheKey);` right before that [cache `getOrSet()` call is made](https://github.com/nystudio107/craft-seomatic/blob/d5552c7/src/services/MetaContainers.php#L191) which cleared the cache on each request, which got me what I needed. I don't suggest that as a permanent fix of course!

**Comment by khalwat** [comment]
Well, but the thing is... it should be fine that the meta containers that it is loading in are cached; the idea is that it loads in the meta containers before your Twig code executes, and then you can modify them as you see fit... and if the second level cache is disabled, it'll render out all of the results.

**Comment by benjamin-smith** [comment]
Not sure what the deal is, I'll have to keep investigating. In `MetaTagContainer::includeMetaData()` I am putting a `die()` when there is a cache miss. It does not fire, even though I am running those cache clearing Twig tags which should trigger the `TagDependency::invalidate()` call a few lines higher up.

```php
    public function includeMetaData($dependency)
    {
        Craft::beginProfile('MetaTagContainer::includeMetaData', __METHOD__);
        $uniqueKey = $this->handle.$this->handle.$dependency->tags[3];
        $cache = Craft::$app->getCache();
        if ($this->clearCache) {
            TagDependency::invalidate($cache, $this::CONTAINER_TYPE.$uniqueKey);
        }
        $tagData = $cache->getOrSet(
            $this::CONTAINER_TYPE.$uniqueKey,
            function () use ($uniqueKey) {
                Craft::info(
                    $this::CONTAINER_TYPE.' cache miss: '.$uniqueKey,
                    __METHOD__
                );
                die('cache miss'); // <===========
```

**Comment by khalwat** [comment]
Okay, I apologize -- I jumped the gun earlier, the change I made is incorrect. I was conflating the cache key with the dependency tags.

I've reverted that push, if you want to pull from `dev-develop`: https://github.com/nystudio107/craft-seomatic/commit/76eb621e5149e1c1f5e323d69711c4a8a820f29e

TL;DR is that the cache `key` is not the same as the cache dependency tags that we use to clear the caches.

If you revert this back, your `die()` should be getting hit, because we're telling it to clear the cache for any tag dependency that matches `$uniqueKey` -- which is one of the tags passed in via `$dependency`

**Comment by khalwat** [comment]
Okay, I have it sorted @benjamin-smith -- it's similar to what I described, but backwards. :)

It's doing:

```
        $uniqueKey = $this->handle.$dependency->tags[3];
```
And then trying to clear that dependency via:

```
        if ($this->clearCache) {
            TagDependency::invalidate($cache, $uniqueKey);
        }
```

The problem is, the what it's trying to invalidate is _not_ one of the tag dependencies because it has `$this->handle` prepended to it.

So... once more, with feeling: https://github.com/nystudio107/craft-seomatic/commit/e8298c14b398d2df5e8f9e61f16d6bd074d17f2e

Pull from `dev-develop` and away we go!

**Comment by benjamin-smith** [comment]
Thank you! Please do not apologize. I'll check this tomorrow as I left my Mac with the source code at the office.

**Comment by benjamin-smith** [comment]
Working over here! The cache clearing Twig tags are indeed clearing the caches, and I am able to see the robots tag in the final HTML output as I expect. Thanks for all of your help with this!

**Comment by khalwat** [comment]
Awesome, it'll be in the next release. Sorry for the run-around yesterday!
